autorecsys/pipeline/preprocessor.py

```diff
@@ -78,6 +103,216 @@ def preprocessing(self, **kwargs):
+                    categ_count_dict[i][line[j]] += 1  # count category occurrences
+
+        # Step 2: Create dictionary for indexing categories.
+        if Path(self.fit_dictionary_path).is_file():
+            categ_index_dict = load_pickle(self.fit_dictionary_path)
+        else:
+            # Define defaultdict where key="feature index" & val="defaultdict" where key="category" & val="index".
+            categ_index_dict = defaultdict(lambda: defaultdict(int))
+
+            for feat_index, categ_dict in categ_count_dict.items():
+                categ_index = 0
+                for categ, count in categ_dict.items():
+                    if count >= self.categ_filter:  # index filtered categories at a later stage
+                        categ_index_dict[feat_index][categ] = categ_index
+                        categ_index += 1
+            del categ_count_dict  # release memory
+            gc.collect()
+            # Save fit dictionary.
+            save_pickle(self.fit_dictionary_path, dict(categ_index_dict))  # cannot pickle defaultdict using lambda
+
+        # Step 3: Index categories.
+        for i, categ_l in enumerate(categ_ll):
+            for j, c in enumerate(categ_l):
+                if c in categ_index_dict[i]:
+                    categ_ll[i][j] = categ_index_dict[i][c]  # index in-place
+                else:  # index filtered categories as the end index
+                    categ_ll[i][j] = len(categ_index_dict[i])
+        del categ_index_dict  # release memory
+        gc.collect()
+
+        # Step 4: Format data and obtain hash statistics.
+        array_data = np.concatenate((np.asarray(label_ll).T, np.asarray(numer_ll).T, np.asarray(categ_ll).T), axis=1)
+        del label_ll  # release memory
+        del numer_ll
+        del categ_ll
+        gc.collect()
+
+        self.pd_data = pd.DataFrame(array_data, columns=self.used_columns_names)
+        self.hash_sizes = [self.pd_data[categ_name].nunique() for categ_name in self.categ_names]
+        del array_data  # release memory
+        gc.collect()
+
+        for col_name, dtype in self.dtype_dict.items():
+            self.pd_data[col_name] = self.pd_data[col_name].astype(dtype)
+        # Save transformed data.
+        save_pickle(self.transformed_data_path, self.pd_data)
+
+    def scale_numerical_data(self):
+
+        # TODO: Designated transformation functions to specified place.
+        def scale_by_natural_log(num):
+            # TODO: 1) Explain why the conditional statement makes exception for numbers like 1, where 1>ln(1)**2
+            if num > 2:
+                num = math.log(float(num))**2
+            return num
+
+        for numer_name in self.numer_names:
+            self.pd_data[numer_name] = self.pd_data[numer_name].map(scale_by_natural_log)
+
+    def split_data(self, X, y, train_size, validate_size, random_state=42):
+        """
+
+        :param X: numpy array
```

Comment:
Follow the docstirng style of other files to modify it after you finish other functionalities.

Reply:
Agreed